### DISCUSSION TITLE
Clarification on Default Language Extensions

### DISCUSSION THREAD

**Discussion by katrynaz** [CodinGame/monaco-vscode-api#339] [General]
Hello, I have a couple of questions regarding the default language extensions.  
- Do the imports need to be in any particular order when initializing?
- Are there any other required imports to get the languages to work? 

Since `monaco-editor-core` doesn't offer the languages out of the box, I've opted to manually add the language support in, but `monaco.languages.getLanguages()` still only seems to detect PlainText. 

This is the import I am using: 
```jsx
import '@codingame/monaco-vscode-python-default-extension';
```

Further, this import seems to give me a bunch of console warnings: 
```jsx
import "@codingame/monaco-vscode-typescript-language-features-default-extension";
```
<img width="825" alt="Screenshot 2024-02-07 at 9 17 43 AM" src="https://github.com/CodinGame/monaco-vscode-api/assets/98342633/176c03e2-e4bb-443a-8820-be554829f0cd">

Please let me know if I'm missing anything, I've read through the documentation multiple times and couldn't find a solution. 
I've also seen similar discussions here, which were able to get me further, but still didn't resolve my issue.  

I'm using Vite in an Electron application and setup seems to be working fine.  Any help is appreciated!

**Comment by CGNonofr** [comment]
Hello

> Do the imports need to be in any particular order when initializing?

Not at all, you can even import it dynamically at any time

> Are there any other required imports to get the languages to work?

In VSCode there is a notion of `extension points` which correspond to a field in the package.json `contributes` field. In the VSCode extensions that add languages, at least 2 of them are required: `languages` and `grammars`

Those extension points are only available if the corresponding service override is registered:
- languages => language service override
- grammars => textmate service override

> Further, this import seems to give me a bunch of console warnings: [...]

Never seen that error before! Are you using the last version?

> Please let me know if I'm missing anything, I've read through the documentation multiple times and couldn't find a solution.
I've also seen similar discussions here, which were able to get me further, but still didn't resolve my issue

A minimal reproduction repo would probably help :)

> I'm using Vite in an Electron application and setup seems to be working fine. Any help is appreciated!

Keep in mind that this library bundles the part of VSCode designed for its web version so it may be not optimal or adapted to an electron app

**Reply by katrynaz** [reply]
Thanks for your quick response

> Never seen that error before! Are you using the last version?

I am using the latest as far as yesterday (removed `node_modules` and reinstalled)

I'll attempt these suggestions and get latest from today.  I'll add a sample of the repo shortly

**Comment by katrynaz** [comment]
@CGNonofr Hello, I have added the suggestions from the previous response but was still unsuccessful in seeing any of the manually added default extensions.  The packages have been upgraded to the latest as of today (resulted in v^2.0.3)

Update to `package.json`: 
```json
"contributes": {
    "languages": [
      {
        "id": "python",
        "extensions": [
          ".py"
        ],
        "aliases": [
          "Python",
          "python"
        ],
        "configuration": "./node_modules/@codingame/monaco-vscode-python-default-extension/resources/language-configuration.json"
      }
    ],
    "grammars": [
      {
        "language": "python",
        "scopeName": "source.python",
        "path": "./node_modules/@codingame/monaco-vscode-python-default-extension/resources/MagicPython.tmLanguage.json"
      }
    ]
  }
  ```

I've created a minimal reproduction of the repo found here: 
https://github.com/katrynaz/monaco-vscode-lite/blob/main/src/monaco-setup/index.ts 

> Keep in mind that this library bundles the part of VSCode designed for its web version so it may be not optimal or adapted to an electron app

Indeed, there have been things I've had to "hack" to accommodate the application configuration. 

Thanks

**Reply by CGNonofr** [reply]
Sorry if it was unclear, you don't have to add the contributes section to your own package.json, it's defined in the VSCode extension (You have nothing to do, I was just explaining how it works)

The issue is probably that you didn't register the textmate service override

**Reply by Yokozuna59** [reply]
Hi @CGNonofr, Could you please provide a minimal code to be able to do so?

My TextMate grammar has multiple embedded languages, so set the `configuration`,`language`, and `grammar` for them.

**Reply by CGNonofr** [reply]
As I previously said, you need the grammar extension point.

There is also a lot of wrong things. so many I've preferred opening a PR => https://github.com/katrynaz/monaco-vscode-lite/pull/1
